pkg/types/tuf/v0.0.1/entry.go

```diff
@@ -0,0 +1,406 @@
+/*
+Copyright © 2021 The Sigstore Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package tuf
+
+import (
+	"context"
+	"crypto/sha256"
+	"crypto/sha512"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/theupdateframework/go-tuf/data"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/sigstore/rekor/pkg/log"
+	"github.com/sigstore/rekor/pkg/types"
+	"github.com/sigstore/rekor/pkg/types/tuf"
+	"github.com/sigstore/rekor/pkg/util"
+
+	"github.com/go-openapi/strfmt"
+
+	"github.com/sigstore/rekor/pkg/pki"
+
+	"github.com/go-openapi/swag"
+	"github.com/sigstore/rekor/pkg/generated/models"
+)
+
+const (
+	APIVERSION = "0.0.1"
+)
+
+func init() {
+	if err := tuf.VersionMap.SetEntryFactory(APIVERSION, NewEntry); err != nil {
+		log.Logger.Panic(err)
+	}
+}
+
+type BaseSigned struct {
+	Type    string    `json:"_type"`
+	Expires time.Time `json:"expires"`
+	Version int       `json:"version"`
+}
+
+type V001Entry struct {
+	TufObj                  models.TufV001Schema
+	fetchedExternalEntities bool
+	keyObj                  pki.PublicKey
+	sigObj                  pki.Signature
+}
+
+func (v V001Entry) APIVersion() string {
+	return APIVERSION
+}
+
+func NewEntry() types.EntryImpl {
+	return &V001Entry{}
+}
+
+func (v V001Entry) IndexKeys() []string {
+	var result []string
+
+	if v.HasExternalEntities() {
+		if err := v.FetchExternalEntities(context.Background()); err != nil {
+			log.Logger.Error(err)
+			return result
+		}
+	}
+
+	// Index manifest hash, type, and version.
+	manifestHash := sha512.Sum512([]byte(v.TufObj.Manifest.Signed.Content))
+	result = append(result, strings.ToLower(hex.EncodeToString(manifestHash[:])))
+	result = append(result, v.TufObj.Manifest.Type)
+	result = append(result, v.TufObj.Version)
+
+	// Index root.json hash.
+	root, err := v.keyObj.CanonicalValue()
+	if err != nil {
+		log.Logger.Error(err)
+	} else {
+		rootHash := sha256.Sum256(root)
+		result = append(result, strings.ToLower(hex.EncodeToString(rootHash[:])))
+	}
+
+	// TODO: Index individual key IDs?
```

Comment:
Aha, I was looking for this :) 

Yes, I think it would be super useful to be able to query Rekor for TUF metadata based on keyids. Could we pretty please add this? Happy to help with this effort also.

Reply:
Yes!! Let's do this in a follow up. Should be pretty easy to do by adding a method for KeyIDs in the PublicKey type for TUF